dill/_dill.py

```diff
@@ -1176,7 +1211,7 @@ def _save_with_postproc(pickler, reduction, is_pickler_dill=None, obj=Getattr.NO
 # License: https://github.com/cloudpipe/cloudpickle/blob/master/LICENSE
 @register(CodeType)
 def save_code(pickler, obj):
-    log.info("Co: %s" % obj)
+    pickler._log_trace("Co: %s", obj)
```

Comment:
Because of limitations from long ago (#456), the pickler that is passed into the registered save functions is not necessarily a dill pickler and could have a missing `_log_trace` function.

I think that formatting the log trace as a tree is a good idea, but I think you should be using a `Formatter` instead of the pickler itself to do this:

https://stackoverflow.com/questions/35048921/format-log-messages-as-a-tree

Reply:
Oh, now I get the whole `is_dill()` thing. It is related to other packages that also extend `pickle`. I'll take a look at `Formatter` then.

Reply:
Other people can use the default pickler `pickle.Pickler` instead of `dill.Pickler` and they shouldn't see an error. This system is like this because in older versions of Python, the pickler class couldn't have its own dispatch. I think that it would be quick change to make the dispatch per class instead of global (cloudpickle did it), but it may be a breaking change and I didn't find time to verify that it isn't.

Reply:
I think it's not so hard to verify whether the package is working after such a change, but for code that _extends_ dill it could be tricky. There are both benefits and drawbacks of being able to extend pickle just by importing dill. Maybe it could be made an optional feature somehow.

Hey, look at this! My notes from learning how to use `logging.Formatter` (it's a lot of fun):

```
Basic calling sequence with StreamHandler and Formatter in Python 3.7: 
(All methods from the second line on are from logging.__init__.py)     
                                                                       
dill.logging.logger.log ->   (or info, warning, etc.)                  
  Logger.log ->                                                        
    Logger._log ->                                                     
      Logger.makeRecord ->                                             
        LogRecord.__init__                                             
      Logger.handle ->                                                 
        Logger.callHandlers ->                                         
          Handler.handle ->                                            
            StreamHandler.emit ->                                      
              Handler.format ->                                        
                Formatter.format ->                                    
                  LogRecord.getMessage                                 
                  Formatter.formatMessage ->                           
                    PercentStyle.format                                
```